File: src/org/cerion/symcalc/expression/ErrorExpr.java
package org.cerion.symcalc.expression;

public class ErrorExpr extends Expr {

	public void show(int i) 
	{
		indent(i,"Error " + this.toString());
	}

	@Override
	protected Expr evaluate() {
		return this; 
	}

	@Override
	public boolean equals(Expr e) {
		// There should never be a reason to compare errors so just say they are never equal
		return false;
	}

	public String value()
	{
		return (String)getValue();
	}
	
	public ErrorExpr(String error) {
		setValue(error);
	}

	public static ErrorExpr getInvalidParameterType(Class c, int position) {
		String error = String.format("parameter {%d} must be a %s", position, c.getSimpleName());
		return new ErrorExpr(error);
	}

	public ExprType getType()
	{ 
		return ExprType.ERROR;
	}

	@Override
	public String toString() 
	{
		return "Error(" + value() + ")";
	}
}

File: src/org/cerion/symcalc/expression/ErrorExpr.kt
package org.cerion.symcalc.expression

class ErrorExpr(error: String) : Expr() {

    override val type: ExprType
        get() = ExprType.ERROR

    init {
        value = error
    }

    override fun treeForm(i: Int) = indent(i, "Error " + this.toString())
    override fun evaluate(): Expr = this
    override fun toString(): String = "Error($value)"

    // There should never be a reason to compare errors so just say they are never equal
    override fun equals(e: Expr): Boolean = false
}
